### ISSUE TITLE
Update Android targetSdkVersion to 29 (Android 10), by 2020-11 deadline

### ISSUE THREAD

**Issue by gnprice** [zulip/zulip-mobile#3665]
The next round after #3563 (and #3075). We should update our `targetSdkVersion` to 29, aka Android 10.

Deadline is 2020-11-02. But we could do this any time (indeed could have as soon as Android 10 was out last year), and in any case filing an issue to give us a place to track tidbits we learn.

* General docs on this upgrade are [here](https://developer.android.com/distribute/best-practices/develop/target-sdk.html#pre10).
* The list of behavior changes triggered by actually bumping the targetSdkVersion is [here](https://developer.android.com/about/versions/10/behavior-changes-10).

The important steps for this upgrade are:
1. Read about the potentially-breaking changes, and identify those that might affect us.
2. Make a WIP change to `targetSdkVersion`.
3. Test the WIP change thoroughly, especially in the areas highlighted in step 1. Fix things as needed, and repeat.

Specific tidbits we've learned:
* While validating the previous upgrade, to API 28 aka Android 9 Pie, @ray-kraesig noticed we are said to use some "weakly-deprecated APIs": https://github.com/zulip/zulip-mobile/issues/3563#issuecomment-546106460
  * Of these, one is... IIUC Android 10 puts it [in the "greylist-max-p" list](https://developer.android.com/about/versions/10/non-sdk-q), which means that a `targetSdkVersion` of 29 or greater prevents using it.

**Comment by gnprice** [comment]
The Play Console has started showing me a warning when I upload a new release; first saw it today with the 27.154 release (now in beta). Following some links, [the deadline is 2020-11-02](https://support.google.com/googleplay/android-developer/answer/113469#targetsdk).

(I'll edit the description with more info.)

**Comment by gnprice** [comment]
Bumped the priority, as this deadline is getting sooner.

**Comment by gnprice** [comment]
> 1. Read about the potentially-breaking changes, and identify those that might affect us.

Here's notes from reading through that [list of behavior changes](https://developer.android.com/about/versions/10/behavior-changes-10) triggered by actually bumping the targetSdkVersion. I intend to update this as I learn more.

* [ ] Privacy changes, which have [their own docs page](https://developer.android.com/about/versions/10/privacy/changes).
* [ ] Updates to non-SDK interface restrictions -- i.e. they've added to the lists of undocumented system interfaces that are forbidden to use.
  * [ ] Go through the list and try to identify anything likely to affect us.
  * [ ] Test the WIP change to try to spot breakage.
* [x] Shared memory. We certainly don't have any code of our own that touches this -- we'd only be affected if something in RN was affected, and if so it'd probably be something quite central that virtually all RN apps relied on. We're using RN v0.62, which came out well after Android 10, so it seems unlikely there's a problem of that form.
* [x] Removed execute permission for app home directory. Same story as shared memory.
* [x] Android runtime only accepts system-generated OAT files. Same story.
* [x] Enforcing AOT correctness in ART. Same story.
* [x] Permissions changes for fullscreen intents. We don't do this.
* [ ] Support for foldables.

**Comment by gnprice** [comment]
OK, I've now completed the steps of this that are about reading through the documentation. The results are expanded out into the checklist in the preceding comment. Almost all those items are checked off, but there are a couple of them remaining that correspond to deeper investigation and/or testing.

Here's a fresh task list, covering those remaining items above and zooming back out to the overall issue as set out in the description:
* [ ] Non-SDK interfaces: Look at what the Google Play Console flags as things we seem to be using.
* Make a WIP change that bumps the targetSdkVersion.
* [ ] Test that change:
  * [ ] Attempt to take a photo, in a fresh install of the app so it doesn't already have the CAMERA permission.
  * [ ] Test thoroughly in general, looking for breakage.
  * Fix things as needed, and repeat.
* Commit and push the change.
* Optional followups:
  * Consider using the [new manifest attribute](https://developer.android.com/about/versions/10/behavior-changes-10#foldables) `android:minAspectRatio`.

**Comment by gnprice** [comment]
> Attempt to take a photo, in a fresh install of the app so it doesn't already have the CAMERA permission.

Hmm, well -- I get this error modal:
![image](https://user-images.githubusercontent.com/28173/96654949-3e811f80-12f1-11eb-8621-79e34fc360d4.png)

Specifically, my steps were:
* Start a debug build, in a fresh install, with `targetSdkVersion` at 29, and with Flipper disabled to avoid [this issue](https://chat.zulip.org/#narrow/stream/243-mobile-team/topic/Flipper/near/1042657) which I discovered when trying to test this.
* Attempt to take a photo.
* Get the two permissions prompts as usual; grant both.
* This error modal appears.

Then:
* Try hitting the "camera" button again.
* The same error modal appears.

So it seems there is some problem. It may not be the `getCameraCharacteristics` problem I anticipated above; that doesn't seem to fit with having gotten successfully through granting camera permissions.

Hmm, thinking some more, ISTR that the library we use for getting images from the camera (`react-native-image-picker`) does something a bit odd where it first has the system take a photo and store it, then it goes and reads the resulting file by name. That sounds like the kind of thing that would be blocked by the change ["External storage access scoped to app files and media"](https://developer.android.com/about/versions/10/privacy/changes#scoped-storage) -- what it's actually doing is perfectly innocent (reading the photo the user just explicitly chose to take for use with Zulip), but it's unnecessarily using APIs that could provide access to arbitrary files and are now blocked. And these symptoms seem like a perfect fit for that hypothesis.

So that's a potential lead for investigation; if that's what's happening, we'll want either a newer version of this library, or some other library if it's effectively unmaintained and has no such version, or some other solution.

**Comment by gnprice** [comment]
Well, here's a sad but adequate workaround for the camera issues:
```diff
--- a/android/app/src/main/AndroidManifest.xml
+++ b/android/app/src/main/AndroidManifest.xml
@@ -25,2 +25,3 @@
-        android:theme="@style/AppTheme">
+        android:theme="@style/AppTheme"
+        android:requestLegacyExternalStorage="true">
         <activity
```

That fixes the trouble mentioned in my previous comment, and also a related issue that appears when trying to pick an existing image.

It's a temporary solution -- effectively [it opts out of this aspect](https://developer.android.com/training/data-storage/use-cases#opt-out-scoped-storage) of the effects of targeting Android 10, and when in the future we target Android 11 the opt-out will be ignored. But it's enough to unblock this for now.

I'll file a separate issue for the follow-up, with what I learned from investigating this today.

**Comment by gnprice** [comment]
After the issues with react-native-image-picker described in the previous two comments (affecting both the take-photo and upload-existing-photo features, behind the "camera" and "photo" icons in the compose menu) where it hasn't adapted to the new "scoped storage", I tested out the other feature that seemed likely to be affected by scoped storage: the upload-file feature, behind the "document" icon.

That one, happily, seems to work just fine when targeting Android 10 aka API 29, without needing the `requestLegacyExternalStorage` opt-out / workaround. That feature has a similar flavor but is provided by a different library -- `react-native-document-picker` -- so I guess the latter library has successfully adapted to the new APIs.
